pkg/controller/elasticsearch/driver/upgrade_deletion_driver.go

```diff
@@ -0,0 +1,185 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/expectations"
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DeletionDriver is a controller that given a context uses a plugable strategy
+// to determine which Pods can be safely deleted.
+type DeletionDriver interface {
+	// Delete goes thought the candidates and actually deletes the ones
+	Delete(candidates []*corev1.Pod) (deletedPods []*corev1.Pod, err error)
+}
+
+// DefaultDeletionDriver is the default deletion driver.
+type DefaultDeletionDriver struct {
+	client       k8s.Client
+	esClient     esclient.Client
+	es           *v1alpha1.Elasticsearch
+	statefulSets sset.StatefulSetList
+	state        ESState
+	healthyPods  map[string]corev1.Pod
+	strategy     DeletionStrategy
+	expectations *expectations.Expectations
+}
+
+// NewDeletionDriver creates a new default deletion driver.
+func NewDeletionDriver(
+	client k8s.Client,
+	esClient esclient.Client,
+	es *v1alpha1.Elasticsearch,
+	statefulSets sset.StatefulSetList,
+	state ESState,
+	masterNodesNames []string,
+	healthyPods map[string]corev1.Pod,
+	podsToUpgrade []corev1.Pod,
+	expectations *expectations.Expectations,
+) *DefaultDeletionDriver {
+	return &DefaultDeletionDriver{
+		client:       client,
+		esClient:     esClient,
+		es:           es,
+		statefulSets: statefulSets,
+		state:        state,
+		healthyPods:  healthyPods,
+		expectations: expectations,
+		strategy:     NewDefaultDeletionStrategy(state, healthyPods, podsToUpgrade, masterNodesNames),
+	}
+}
+
+// Delete runs through a list of potential candidates and select the ones that can be deleted.
+// Do not run this function unless driver expectations are met.
+func (d *DefaultDeletionDriver) Delete(candidates []corev1.Pod) (deletedPods []corev1.Pod, err error) {
+	if len(candidates) == 0 {
+		return deletedPods, nil
+	}
+	es := k8s.ExtractNamespacedName(d.es)
+
+	// Check if we are not over disruption budget
+	// Upscale is done, we should have the required number of Pods
+	statefulSets, err := sset.RetrieveActualStatefulSets(d.client, es)
+	if err != nil {
+		return deletedPods, err
+	}
+	expectedPods := statefulSets.PodNames()
+	unhealthyPods := len(expectedPods) - len(d.healthyPods)
+	maxUnavailable := 1
+	// TODO: use GroupingDefinition
```

Comment:
Re: grouping definition. I think we need to have maybe a discussion what the added benefit of having these grouping definitions is and wether we still need them? cc @nkvoll